README.md
# TechLib Package Management Kit V2020_0410a

## Define the following environment variables in the shell:

	TECHLIB_PKGS	- package source directory (*.tgz and *.tlp)
	TECHLIB_RELN	- release note respository directory (sorted collection)
	TECHLIB_HOME	- target techLib directory (where you plan to install)

Directory:

	$(TECHLIB_HOME)/
		+-- T28HPC
			+----- 0p5
			|	+----- FDK
			|		+----- PDK
			|		+----- CTK
			+----- 0p9
				+----- FDK
				|	+----- PDK
				+----- FIP
					+----- STDCELL

## Step 0：Prepare TechLib Package defintion file:

Example:
	
	::::::::::::::
	STDCELL_lib222_7t_base_e.2.0.tlp
	::::::::::::::
	TLP	FORMAT	1.0
	KIT	NODE	T28HPC
	KIT	PDK	0p5
	KIT	GROUP	FIP
	KIT	TYPE	STDCELL
	KIT	VERSION	7t_base_e.2.0
	KIT	ORIGIN	7t_base_e.1.1
	KIT	TOPDIR	lib222_7t_base_e20
	KIT	SIZE	10000
	KIT     MD5SUM	11ba9bfa12c16459bc242c005c351b6f

	REQUIRE	KIT	T28HPC/0p5/FDK/PDK	TOPDIR	pdk222_r10HF7

	; If the kit size is huge, it could be split to multiple package files
	PACKAGE	FILE	STDCELL_lib222_7t_base_e.2.0-1.tgz  
	PACKAGE	FILE	STDCELL_lib222_7t_base_e.2.0-2.tgz  
	PACKAGE	FILE	STDCELL_lib222_7t_base_e.2.0-3.tgz  

	TLP END 
	; below this line, there are human readable docs which will be ignored by tool

## Step 1: Import TechLib Package to release notes repository:

	% tlp_import [--packageSrcDir $TECHLIB_PKGS] [--releaseNoteDir $TECHLIB_RELN]

	% tlp_import <package>.tlp ...

	1. Search TLP definition file (*.tlp) in TECHLIB_PKGS directory 
	2. Copy the TLP definition file to $TECHLIB_RELN directory as a releaseNote file
	   and categorize the releaseNote files based on the collateral category.

  Example:

	releaseNotes/
		|
		+-- T28HPC
			|
			+----- 0p5
			|	|
			|	+----- FDK
			|		|
			|		+----- PDK
			|		|	+---- aaa.releaseNote
			|		|
			|		+----- CTK
			|			+---- bbb.releaseNote
			+----- 0p9
				|
				+----- FDK
				|	|
				|	+----- PDK
				|		+---- ccc.releaseNote
				+----- FIP
					|
					+----- STDCELL
						+---- ddd.releaseNote
						+---- eee.releaseNote


## Step 2: Install selected packages to techlib directory:

  Usage-1 (Package Name):

	% tlp_install <packageName> ...  ; search the tlp from $TECHLIB_PKGS
	% tlp_install <directoryPath>/<packageName>.tlp ...


  Usage-2 (Bundle File):

	% tlp_install --bundleFile <packageBundleListFile>

	  Search the package definition file and tar kit in $TECHLIB_PKGS and install 
	these package follow the sequence. If any package fail to be installed,
	the process will stop. After fixing the problem, same bundleFile could
	be used, package already installed will be kept as is.

  Example:

	packageName1
	packageName2
	packageName3


  Usage-3 (Interactive):

	% tlp_install [--releaseNoteDir $TECHLIB_RELN] [--selectByCategory NODE/PDK/GROUP/TYPE]

	User can specify kit category and then select the packages.
	The tool will list all release notes of selected category under $TECHLIB_RELN directory:

  Example:

	% tlp_install [--releaseNoteDir $TECHLIB_RELN] --selectByCategory T28HPC/0p9

	INFO: Cateogry - [T28HPC/0p9] 

		1) FDK/PDK/ccc
		2) FIP/STDCELL/ddd
		3) FIP/STDCELL/eee

	QUESTION: Which packages would you like to install? 1 3

	INFO: Installing collateral package "ccc" ..
	INFO: Validating package integrity ...
	INFO: Checking package dependcy ...
	INFO: Unpacking package files "ccc-1.tgz" ...
	INFO: Unpacking package files "ccc-2.tgz" ...

	INFO: Installing collateral package "eee" ..
	...

  Logfle:

	tlp_install.log  =>  detail installation log 
	$TECHLIB_HOME/.tlp_install.summar         => pacakge installation tracking summary
	$TECHLIB_HOME/<NODE>/<PDK>/.tlp_packages  => package list which has been installed


## Step 3: Verify post-installation check (NOT IMPLEMETNTED YET)

	% tlp_check <installed_kit_topdir>    (<NODE>/<PDK>/<GROUP>/<TYPE>/<kit_top_dir>)

	% tlp_check --bundleFile <packageBundleListFile>

	% tlp_check [--targetLibDir $TECHLIB_HOME] --techNode <NODE/PDK>



docs/ChangeLog.txt
2017/10/25 Albert Li 
  Create the repository and checkin sample perl code from Haiyong

2017/11/01 Albert Li
  Add the following directory 
	sample/   - sample empty kit directory
	packages/ - tar file created from kits in sample/
	csh/      - demo code using csh/awk
	run/      - testcase

2017/11/02 Albert Li
  Add gawk in csh/, to resolve old version issue in system
  Add dcm_install.csh/dcm_install.awk

2017/11/04 Albert Li
  Add gawk executable for Cygwin and Linux

2017/11/07 Albert Li
  Add dcm_help to view all documents

2020/04/10 Albert Li
  Change package name
  DCM ==> TLP
  dcm ==> tcp
  ICFDK ==> TECHLIB
  1222.2 ==> T28HPC
  x1r0 ==> 0p5

csh/tlp_pack.awk
#!/usr/bin/gawk -f
function HEADER(message)  { print "\033[34;40m"message"\033[0m" }
function HILITE(message)  { print "\033[1m"message"\033[0m" }
function WARNING(message) { print "\033[34mWARNING: "message"\033[0m" }
function ERROR(message)   { print "\033[31;43mERROR: "message"\033[0m" }
function PRINT(message)   { print "\033[31m"message"\033[0m" }
function DEBUG(message)   { print "\033[35mDEBUG: "message"\033[0m" }

BEGIN {
  print "--------------------------------------------------------"
  print "[tlp_pack]: BEGIN"
  print "--------------------------------------------------------"

  tlp_pack_root = ENVIRON["TLP_PACK_TEMP"]
  if (tlp_pack_root == "") {
      tlp_pack_root = "tempLib"
  }

  tlp_pkgs_cfgs = ENVIRON["TLP_CFGS_DEST"] 
  if (tlp_pkgs_cfgs == "") {
      tlp_pkgs_cfgs = "configs"
  }

  tlp_pkgs_dest = ENVIRON["TLP_PKGS_DEST"] 
  if (tlp_pkgs_dest == "") {
      tlp_pkgs_dest = "packages"
  }

  tlp_pack_option= ENVIRON["TLP_PACK_OPTION"]
  split(tlp_pack_option, tlp_options)
  for (option in tlp_options) {
      if (option == "--verbose") {
         option_verbose = 1
      } else if (option == "--info") {
         option_info = 1
      } else if (option == "--test") {
         option_test = 1
      } else {
      }
  }
  


}
BEGINFILE {
  print "\033[34m"
  HILITE("[tlp_pack]: Processing Collateral Package file '"FILENAME"' ...")
  "basename "FILENAME" .csv" | getline bundleFile
  bundleFile = bundleFile".bundle"
  system("echo -n > "bundleFile)

  tlp_format = ""
  tlp_test_mode = 0
  
  kit_node   = "_"
  kit_pdk    = "_"
  kit_group  = "_"
  kit_type   = "_"
  kit_topdir = "_"
  kit_version = "_"
  kit_origin  = "_"
  kit_location = "_"
  kit_content   = ""
  
  req_kit_SKU = "_"
  req_kit_type = "_"

  tlp_pkgs_total   = 0  
  tlp_pkgs_created = 0
  tlp_pkgs_error   = 0  

}

/^#/ {}
/^TLP\s+PACKAGE\s/	{ tlp_format = $3 }
/^TLP\s+MODE\s+TEST/	{ tlp_test_mode = 1 }
/^TLP\s+END\s/		{ nextfile }

/^KIT\s+NODE\s/      { kit_node = $3 }
/^KIT\s+PDK\s/       { kit_pdk  = $3 }
/^KIT\s+GROUP\s/     { kit_group = $3 }
/^KIT\s+TYPE\s/      { kit_type = $3 }
/^KIT\s+TOPDIR\s/    { kit_topdir = $3 }
/^KIT\s+VERSION\s/   { kit_version = $3 }
/^KIT\s+ORIGIN\s/    { kit_origin = $3 }


/^REQUIRE\s+KIT\s/ {
  if ($2 == "KIT") {
     req_kit_type = $3
     if ($4 == "TOPDIR") {
        req_kit_topdir = $5
     } else if ($4 == "VERSION") {
        req_version = $5
     } else {
        req_kit_type = "_"
     }
  } else {
     req_kit_type = "_"
  }
}

/^(FULL|PATCH|MULTI)\s/ {
  tlp_pkgs_total++

  pack_type = $1
  kit_node = $2
  kit_pdk = $3
  kit_group = $4
  kit_type    = $5
  kit_SKU   = $6
  kit_topdir  = $7
  kit_version = $8
  req_kit_SKU  = $9
  kit_location = $10
  kit_content = $11

  HILITE("["tlp_pkgs_total"]: Packing Kit '"kit_SKU"' ...")
  category_dir = kit_node"/"kit_pdk"/"kit_group"/"kit_type

  system("mkdir -p "tlp_pkgs_cfgs)
  system("mkdir -p "tlp_pkgs_dest)
  kit_tlp = tlp_pkgs_cfgs"/"kit_SKU".tlp"
  kit_tgz = tlp_pkgs_dest"/"kit_SKU".tgz"

     print "TLP\tFORMAT\t1.0"	> kit_tlp
     print "" >> kit_tlp
     print "KIT\tNODE\t"kit_node	>> kit_tlp
     print "KIT\tPDK\t"kit_pdk	>> kit_tlp
     print "KIT\tGROUP\t"kit_group	>> kit_tlp
     print "KIT\tTYPE\t"kit_type	>> kit_tlp
     print "KIT\tTOPDIR\t"kit_topdir	>> kit_tlp
     print "KIT\tVERSION\t"kit_version	>> kit_tlp
     print "KIT\tORIGIN\t"kit_origin	>> kit_tlp
     print "" >> kit_tlp

     if ((req_kit_SKU != "_") && (req_kit_SKU != "-")){
        if (req_kit_SKU != kit_SKU) {
           print "REQUIRE\tSKU\t"req_kit_SKU >> kit_tlp
        }
     }

     if ((req_kit_type != "_") && (req_kit_type != "-")) {
        if ((pack_type == "PATCH") || (req_kit_topdir != kit_topdir)) {
           print "REQUIRE\tKIT\t"req_kit_type"\tTOPDIR\t"req_kit_topdir >> kit_tlp
        }
     }
  tlp_pack_dir = tlp_pack_root"/"category_dir
  system("mkdir -p "tlp_pack_dir"/"kit_topdir)
     
  print "PACKAGE\tTYPE\t"pack_type	>> kit_tlp
  file_type = "FILE"
  if (tlp_test_mode == 1) {
     print "       Create test mode kit directory '"kit_topdir"' ..."
     system("cp "kit_tlp" "tlp_pack_dir"/"kit_topdir"/"kit_SKU".releaseNote")
     system("(cd "tlp_pack_dir"; tar -c -O -z "kit_topdir") > "tlp_pkgs_dest"/"kit_SKU"_test.tgz") 
     print "" >> kit_tlp
     print "PACKAGE\t"file_type"\t"kit_SKU"_test.tgz" >> kit_tlp
     print "" >> kit_tlp
     print "TLP END" >> kit_tlp
  } else if ((kit_location == "")||(kit_location == "-")||(kit_location == "_")) {
     print "       Create TLP file only ..."
     print "" >> kit_tlp
     print "PACKAGE\t"file_type"\t"kit_SKU".tgz" >> kit_tlp
     print "" >> kit_tlp
     print "TLP END" >> kit_tlp
  } else if (pack_type == "MULTI") {
     "ls -1 "kit_location"/*.tgz" | getline pkgs_files
     print "" >> kit_tlp
     print "PACKAGE\t"file_type"\t"pkgs_files >> kit_tlp
     print "" >> kit_tlp
     print "TLP END" >> kit_tlp
  } else if (system("test -f "kit_location) == 0) {
     print "       Use package file '"kit_location"' ..."
     print "" >> kit_tlp
     print "PACKAGE\t"file_type"\t"kit_location >> kit_tlp
     print "" >> kit_tlp
     print "TLP END" >> kit_tlp
  } else if (system("test -d "kit_location) == 0) {
     if (kit_content == "") { kit_content = "." }
     "basename "kit_location"" | getline kit_dir_base
     if (kit_dir_base == kit_topdir) {
        print "       Create package file '"kit_SKU".tgz' .."
        #print "(cd "kit_location"/..; tar -c -O -z "kit_topdir") > "kit_tgz
        system("(cd "kit_location"/..; tar -c -O -z "kit_topdir") > "kit_tgz) 
     } else {
        print "       Copy collateral to kit directory '"kit_topdir"' ..."
        #print "(cd "kit_location"; tar -c -O "kit_content" )|( cd "tlp_pack_dir"/"kit_topdir"; tar -x )"
        system("(cd "kit_location"; tar -c -O "kit_content" )|( cd "tlp_pack_dir"/"kit_topdir"; tar -x )")
        print "       Create package file '"kit_SKU".tgz' .."
        #print "(cd "tlp_pack_dir"; tar -c -O -z "kit_topdir") > "kit_tgz
        system("(cd "tlp_pack_dir"; tar -c -O -z "kit_topdir") > "kit_tgz) 
     }
     print "" >> kit_tlp
     print "PACKAGE\t"file_type"\t"kit_SKU".tgz" >> kit_tlp
     print "" >> kit_tlp
     print "TLP END" >> kit_tlp
     "md5sum "kit_tgz | getline tgz_md5sum
  } else {
     ERROR("Kit directory '"kit_location"' does not exist.")
     system("rm -fr "tlp_pack_dir"/"kit_topdir)
     tlp_pkgs_error++
  }
  if (system("test -d "tlp_pack_dir"/"kit_topdir) == 0 ) {
     print kit_SKU > bundleFile
     system("rm -fr "tlp_pack_dir"/"kit_topdir)
     tlp_pkgs_created++
  }
}

ENDFILE {
}
END {
  print "\033[1m\033[34m"
  print "--------------------------------------------------------"
  print "[tlp_pack]: Total "tlp_pkgs_created"/"tlp_pkgs_total" tlp packages are created." 
  if (tlp_pkgs_error) {
  print "[tlp_pack]: Total "tlp_pkgs_error"/"tlp_pkgs_total" tlp files have missing package files. (error)"
  }
  print "--------------------------------------------------------"
  print "\033[0m"
}

csh/tlp_import.awk
#!/usr/bin/gawk -f
function HEADER(message)  { print "\033[34;40m"message"\033[0m" }
function HILITE(message)  { print "\033[1m"message"\033[0m" }
function WARNING(message) { print "\033[34mWARNING: "message"\033[0m" }
function ERROR(message)   { print "\033[31;43mERROR: "message"\033[0m" }
function PRINT(message)   { print "\033[31m"message"\033[0m" }
function DEBUG(message)   { print "\033[35mDEBUG: "message"\033[0m" }
function find_tlp_package(fname) {
  pkgs_file=""
#  tlp_pkgs_source = ENVIRON["TECHLIB_PKGS"] 
  split(tlp_pkgs_source, pkgs_path, ":")
  for (i in pkgs_path) {
      pkgs_file=pkgs_path[i]"/"fname
#      DEBUG("Checking file :"pkgs_file)
      if (system("test -e "pkgs_file)==0) {
#         DEBUG("Package Found :"pkgs_file)
         return pkgs_file
      }
  }
#  DEBUG("Package Not Found :"fname)
  return ""
}

BEGIN {
  print "--------------------------------------------------------"
  print "[tlp_import]: BEGIN "
  print "--------------------------------------------------------"
  tlp_pkgs_source = ENVIRON["TECHLIB_PKGS"] 
  if (tlp_pkgs_source == "") {
      tlp_pkgs_source = "."
  }

  tlp_pkgs_config = ENVIRON["TECHLIB_CFGS"] 
  if (tlp_pkgs_config == "") {
      tlp_pkgs_config = "."
  }
  
  tlp_reln_root = ENVIRON["TECHLIB_RELN"] 
  if (tlp_reln_root == "") {
      tlp_reln_root = "releaseNotes"
  }
  tlp_install_root = ENVIRON["TECHLIB_HOME"]
  if (tlp_install_root == "") {
      tlp_install_root = "techLib"
  }

  tlp_import_option= ENVIRON["TLP_PACK_OPTION"]
  split(tlp_import_option, tlp_options)
  for (option in tlp_options) {
      if (option == "--verbose") {
         mode_verbose = 1
      } else if (option == "--info") {
         mode_info = 1
      } else if (option == "--skip_topdir") {
         skip_topdir = 1
      } else {
      }
  }
    
  tlp_totoal   = 0
  tlp_created   = 0
  tlp_skipped  = 0
  tlp_modified = 0
  tlp_pkgs_err = 0
}

BEGINFILE {
  tlp_total++
  HILITE("["tlp_total"]: Reading '"FILENAME"' ...")
  tlp_format    = "1.0"
  kit_node      = _
  kit_pdk       = _
  kit_group     = _
  kit_type      = _
  kit_version   = _
  kit_orgin     = _
  kit_topdir    = _
  kit_size      = _
  kit_md5sum    = _

  base_num      = 0
  file_num      = 0
  pkgs_file_missing = 0
}
/^#/ { next }
/^TLP\s+FORMAT\s/	{ tlp_format = $3 }
/^TLP\s+END\s/		{ nextfile }

/^KIT\s+NODE\s/      { kit_node = $3 }
/^KIT\s+PDK\s/       { kit_pdk  = $3 }
/^KIT\s+GROUP\s/     { kit_group = $3 }
/^KIT\s+TYPE\s/      { kit_type = $3 }
/^KIT\s+TOPDIR\s/    { kit_topdir = $3 }
/^KIT\s+VERSION\s/   { kit_version = $3 }
/^KIT\s+ORIGIN\s/    { kit_origin = $3 }
/^KIT\s+SIZE\s/      { kit_size = $3 }


/^PACKAGE\s+TARGET\s/ { 
  tlp_package_dir = $3 
}
/^PACKAGE\s+BASE\s/  {
    base_num++ 
    pkgs_base[base_num]=$3
    tlp_pkgs_base = tlp_pkgs_config"/"pkgs_base[base_num]".tlp"
    if (system("test -e "tlp_pkgs_base) != 0) {
       pkgs_file_missing++
       PRINT("    : Install base - "pkgs_base[base_num]" can not be found in package source.")
    } else {
       print "    : Install base - "tlp_pkgs_base
    }
}
/^PACKAGE\s+(FILE|PATCH)\s/  {
    file_num++
    file_lst[file_num]=$3
    file_top[file_num]=$4
    file_md5[file_num]=$5
    file_tgz[file_num]=find_tlp_package($3)
    if (file_tgz[file_num] == "") {
       pkgs_file_missing++
       PRINT("    : Pacakge file - "file_lst[file_num]" can not be found in package source.")
    } else {
       print "    : Package file - "file_tgz[file_num]
    }
}


ENDFILE {
  if (verbose_mode == 1) {
     print "TLP FORMAT    "tlp_format
     print "KIT NODE      "kit_node
     print "KIT PDK       "kit_pdk
     print "KIT GROUP     "kit_group
     print "KIT TYPE      "kit_type
     print "KIT VERSION   "kit_version
     print "KIT ORIGIN    "kit_origin
     print "KIT TOPDIR    "kit_topdir
     print "KIT MD5SUM    "kit_md5sum
     print "TLP END"
  }
  if (skip_topdir == 1) {
     reln_dir = kit_node"/"kit_pdk"/"kit_group"/"kit_type
  } else {
     reln_dir = kit_node"/"kit_pdk"/"kit_group"/"kit_type"/"kit_topdir
  }
  "basename "FILENAME" .tlp" | getline tlp_basename
  reln_file = reln_dir"/"tlp_basename".releaseNote"

  if (pkgs_file_missing) {
     ERROR("Kit '"tlp_basename"' has "pkgs_file_missing" missing pacakge files")
     tlp_pkgs_err++
  } else if (system("test -e "tlp_reln_root"/"reln_file) == 0) {
     WARNING("Skip import "reln_file" (already exist)")
     if (system("diff "tlp_reln_root"/"reln_file" "FILENAME) !=0) {
        WARNING("Kit '"tlp_basename"' in the DK_RELN has been modified")
        tlp_modified++
     }
     tlp_skipped++
  } else {
     print "    : Creating '"reln_file"' ("kit_version")"
     system("mkdir -p "tlp_reln_root"/"reln_dir)
     system("cp -f "FILENAME" "tlp_reln_root"/"reln_file)
     tlp_created++
     print kit_node" "kit_pdk" "kit_group" "kit_type"\t"tlp_basename" "kit_topdir" "kit_version" "kit_origin >> tlp_reln_root"/.tlp_package_info.csv"
  }
  print ""
}

END {
  print "\033[1m\033[34m"
  print "------------------------------------------------------------------"
  print "[tlp_import]: Total "tlp_created"/"tlp_total" tlp release notes are created." 
  if (tlp_pkgs_err) {
  print "[tlp_import]: Total "tlp_pkgs_err"/"tlp_total" tlp files have missing package files. (error)"
  }
  if (tlp_skipped) {
  print "[tlp_import]: Total "tlp_skipped"/"tlp_total" tlp files are skipped. (exist)"
  }
  if (tlp_modified) {
  print "[tlp_import]: Total "tlp_modified"/"tlp_skipped" existing tlp files are modified.(warning)"
  }
  print "------------------------------------------------------------------"
  print "\033[0m"
}


csh/tlp_install.awk
#!/usr/bin/gawk -f
function HEADER(message)  { print "\033[34;40m"message"\033[0m" }
function HILITE(message)  { print "\033[1m"message"\033[0m" }
function WARNING(message) { print "\033[34mWARNING: "message"\033[0m" }
function ERROR(message)   { print "\033[31;43mERROR: "message"\033[0m" }
function PRINT(message)   { print "\033[31m"message"\033[0m" }
function DEBUG(message)   { print "\033[35mDEBUG: "message"\033[0m" }
function find_tlp_package(fname) {
  pkgs_file=""
  tlp_pkgs_source = ENVIRON["TECHLIB_PKGS"] 
  split(tlp_pkgs_source, pkgs_path, ":")
  for (i in pkgs_path) {
      pkgs_file=pkgs_path[i]"/"fname
#      DEBUG("Checking file :"pkgs_file)
      if (system("test -e "pkgs_file)==0) {
#         DEBUG("Package Found :"pkgs_file)
         return pkgs_file
      }
  }
#  DEBUG("Package Not Found :"fname)
  return ""
}

BEGIN {
  print "--------------------------------------------------------"
  print "[tlp_install]: BEGIN "
  print "--------------------------------------------------------"
  tlp_install_root = ENVIRON["TECHLIB_HOME"]
  if (tlp_install_root == "") {
      tlp_install_root = "techLib"
  }
  tlp_install_summary = tlp_install_root"/.tlp_install.summary"
  if ((system("mkdir -p "tlp_install_root"/.tlp_install/") != 0) ||
      (system("touch "tlp_install_summary) != 0)) {
      ERROR("no write permission on '"tlp_install_root"'.")
      exit -1
  }
  
  tlp_pkgs_source = ENVIRON["TECHLIB_PKGS"] 
  if (tlp_pkgs_source == "") {
      tlp_pkgs_source = "."
  }

  tlp_pkgs_config = ENVIRON["TECHLIB_CFGS"] 
  if (tlp_pkgs_config == "") {
      tlp_pkgs_config = "."
  }
  
  
  tlp_reln_root = ENVIRON["TECHLIB_RELN"] 
  if (tlp_reln_root == "") {
      tlp_reln_root = "releaseNotes"
  }


  tlp_install_option= ENVIRON["TLP_INSTALL_OPTION"]
  split(tlp_install_option, tlp_options)
  for (option in tlp_options) {
      if (option == "--verbose") {
         mode_verbose = 1
      } else if (option == "--info") {
         mode_info = 1
      } else {
      }
  }
  
  tlp_intall_temp = ENVIRON["TLP_INSTALL_TEMP"]
  
  tlp_totoal   = 0
  tlp_created  = 0
  tlp_skipped  = 0
  tlp_conflict = 0
  tlp_modified = 0
  tlp_reqs_err = 0
  tlp_pkgs_err = 0
  
  color_warn = "\033[33m"
  color_err  = "\033[31m"
  color_off  = "\033[0m"
}

BEGINFILE {
  tlp_total++
  HILITE("["tlp_total"]: Reading '"FILENAME"' ...")
  tlp_format    = "1.0"
  kit_node      = _
  kit_pdk       = _
  kit_group     = _
  kit_type      = _
  kit_version   = _
  kit_orgin     = _
  kit_topdir    = _
  kit_size      = _
  kit_md5sum    = _
  
  tlp_package_type = "FULL"

  kit_already_exist = 0
  kit_require_fail = 0
  
  base_num      = 0
  file_num      = 0
  pkgs_file_missing = 0

}
/^#/ { next }
/^TLP\s+FORMAT\s/	{ tlp_format = $3 }
/^TLP\s+END\s/		{ nextfile }

/^KIT\s+NODE\s/      { kit_node = $3 }
/^KIT\s+PDK\s/       { kit_pdk  = $3 }
/^KIT\s+GROUP\s/     { kit_group = $3 }
/^KIT\s+TYPE\s/      { kit_type = $3 }
/^KIT\s+TOPDIR\s/    { kit_topdir = $3 }
/^KIT\s+VERSION\s/   { kit_version = $3 }
/^KIT\s+ORIGIN\s/    { kit_origin = $3 }
/^KIT\s+SIZE\s/      { kit_size = $3 }
/^KIT\s+MD5SUM\s/    { kit_md5sum = $3 }

/^REQUIRE\s+SKU\s/ {
  req_kit_SKU = $3
  if ((req_kit_SKU != "_") && (req_kit_SKU != "-")) {
     if (system("test -e "tlp_install_root"/.tlp_install/"req_kit_SKU".tlp") != 0) {
        PRINT("    : Required kit SKU '"req_kit_SKU"' has not been installed yet.")
        kit_require_fail++
     }
  }
}

/^REQUIRE\s+DIR\s+(\S+)/ {
  req_kit_dir = $3
  if ((req_kit_dir != "_") && (req_kit_dir != "-")) {
     if (system("test -e "tlp_install_root"/"req_kit_dir) != 0) {
        PRINT("    : Required kit dir '"req_kit_dir"' has not been installed yet.")
        kit_require_fail++
     }
  }
}


/^REQUIRE\s+KIT\s/ {
  req_kit_type = $3
  if (req_kit_SKU != "_") {
     if ($4 == "TOPDIR") {
        req_kit_dir = req_kit_type"/"$5
        if (system("test -e "tlp_install_root"/"req_kit_dir) != 0) {
           PRINT("    : Required kit dir '"req_kit_dir"' has not been installed yet.")
           kit_require_fail++
        }
     }
  }
}

/^PACKAGE\s+TYPE\s/ { 
  tlp_package_type = $3 
  if ($3 != "FULL") {
       print "    : Package type - "tlp_package_type
  }
}
/^PACKAGE\s+TARGET\s/ { 
  tlp_package_dir = $3 
}
/^PACKAGE\s+BASE\s/  {
    base_num++ 
    pkgs_base[base_num]=$3
    tlp_pkgs_base = tlp_pkgs_config"/"pkgs_base[base_num]".tlp"
    if (system("test -e "tlp_pkgs_base) != 0) {
       pkgs_file_missing++
       PRINT("    : Install base - "pkgs_base[base_num]" can not be found in package source.")
    } else {
       print "    : Install base - "tlp_pkgs_base
    }
}
/^PACKAGE\s+(FILE|FULL|MULTI|PATCH)\s/  {
    file_num++
    file_pack[file_name]=$2
    file_lst[file_num]=$3
    file_top[file_num]=$4
    file_md5[file_num]=$5
    file_tgz[file_num]=find_tlp_package($3)
    if (file_tgz[file_num] == "") {
       pkgs_file_missing++
       PRINT("    : Pacakge file - "file_lst[file_num]" can not be found in package source.")
    } else {
       print "    : Package file - "file_tgz[file_num]
    }
}


ENDFILE {
  if (verbose_mode == 1) {
     print "TLP FORMAT    "tlp_format
     print "KIT NODE      "kit_node
     print "KIT PDK       "kit_pdk
     print "KIT GROUP     "kit_group
     print "KIT TYPE      "kit_type
     print "KIT VERSION   "kit_version
     print "KIT ORIGIN    "kit_origin
     print "KIT TOPDIR    "kit_topdir
     print "KIT SIZE      "kit_size
     print "KIT MD5SUM    "kit_md5sum
  }
  process_dir = kit_node"/"kit_pdk
  category_dir = kit_node"/"kit_pdk"/"kit_group"/"kit_type
  "basename "FILENAME" .releaseNote" | getline BASEFILE
  "basename "BASEFILE" .tlp" | getline kit_SKU
  tlp_install_dir = tlp_install_root"/"category_dir

  if (system("test -e "tlp_install_dir"/"kit_topdir) == 0) {
     if (system("test -f "tlp_install_dir"/"kit_topdir"/"kit_SKU".releaseNote") == 0) {
        WARNING("Skip install '"kit_SKU"' (already installed)")
        tlp_skipped++
        kit_already_exist = 1
     } else if (tlp_package_type == "FULL") {
        ERROR("Kit directory '"kit_topdir"' already exist before installing full kit package.")
        tlp_conflict++
        kit_already_exist = 1
     } else {
        print "    : Directory '"tlp_install_dir"/"kit_topdir"' already exist."
     }
  } else {
     if (tlp_package_type == "PATCH") {
        WARNING("Base directory '"kit_topdir"' is missing for patch package.")
     }
  }
  if (kit_already_exist) {
  } else if (kit_require_fail) {
     ERROR("Can not install '"kit_SKU"' (dependency fail)")
     tlp_reqs_err++
  } else if (pkgs_file_missing) {
     ERROR("Can not install '"kit_SKU"' (missing pacakge)")
     tlp_pkgs_err++
  } else {
     system("mkdir -p "tlp_install_dir"/"kit_topdir)
     system("cp -f "FILENAME" "tlp_install_dir"/"kit_topdir"/"kit_SKU".releaseNote")
     system("ln -fs ../"category_dir"/"kit_topdir"/"kit_SKU".releaseNote "tlp_install_root"/.tlp_install/"kit_SKU".tlp")
     system("echo \"`date +%Y%m%d_%H%M%S` `whoami` % tlp_install "kit_SKU"\t;"FILENAME"\" >> "tlp_install_summary)
     if (i in pkgs_base) {
         tlp_pkgs_base = tlp_pkgs_config"/"pkgs_base[i]".tlp"
         print "INFO: Installing base kit '"tlp_pkgs_base"' ..."
         system("tlp_install "tlp_pkgs_base)
     }  
     for (i in file_tgz) {
        tlp_pkgs_file = file_tgz[i]
        print "INFO: Unpacking file '"tlp_pkgs_file"' ..."
        system("gunzip -c "tlp_pkgs_file" | (cd "tlp_install_dir"; tar xvf -)")
     }
     print kit_SKU >> tlp_install_root"/"process_dir"/.tlp_packages"
     tlp_created++
  }
  print ""
}

END {
  print "\033[1m\033[34m"
  print "------------------------------------------------------------------"
  if (tlp_created) {
  print "[tlp_install]: Total "tlp_created"/"tlp_total" kits are installed."
  }
  if (tlp_skipped) {
  print "[tlp_install]: Total "tlp_skipped"/"tlp_total" kits are skipped. (already exist)"
  }
  if (tlp_conflict) {
  print "[tlp_install]: Total "tlp_conflict"/"tlp_total" kits have conflict topdir. (error)"
  }
  if (tlp_reqs_err) {
  print "[tlp_install]: Total "tlp_reqs_err"/"tlp_total" kits require check fail. (error)"
  }
  if (tlp_pkgs_err) {
  print "[tlp_install]: Total "tlp_pkgs_err"/"tlp_total" kits have missing package files. (error)"
  }
  print "------------------------------------------------------------------"
  print "\033[0m"
}
